In [1]:
! pip install pysam

     |████████████████████████████████| 10.2 MB 15.3 MB/s eta 0:00:01


In [2]:
import pysam

In [18]:
samfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

## Inspecting first read

In [19]:
for read in samfile:
    print(read)
    print()
    print(f"Read name is {read.qname}")
    print(f"Read flag is {read.flag}")
    print(f"Read reference sequence name is {read.rname}")
    print(f"Postion of read is {read.pos}")
    print(f"Mapping quality is {read.mapq}")
    print(f"Cigar string is {read.cigarstring}")
    print(f"Mate reference sequence name is {read.mrnm}")
    print(f"Mate left-most position is {read.mpos}")
    print(f"Inferred insert size is {read.infer_query_length()}")
    print(f"Query sequence is {read.seq}")
    print(f"Quality sequence  {read.qual}")
    break

C0HVYACXX120402:7:1207:5722:57044	1187	20	9483248	27	76M	20	9483381	76	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]

Read name is C0HVYACXX120402:7:1207:5722:57044
Read flag is 1187
Read reference sequence name is 20
Postion of read is 9483248
Mapping quality is 27
Cigar string is 76M
Mate reference sequence name is 20
Mate left-most position is 9483381
Inferre

<br/>
<b> Flag for this read is 1187 = 0100 1010 0011b </b> <br/>
1 - the read is paired in the sequencing, no metter whether it is mapped in pair <br/>
1 - the read is mapped in a proper pair <br/>
0 - not unmapped <br/>
0 - mate is not unamapped <br/>
0 - foward stand <br/>
1 - mate strand is negative <br/>
0 - the read is not first read in pair <br/>
1 - the read is second read in pair <br/>
0 - the aligment is primary
0 - the read not fails platform/vendor quality check
1 - the read is either a PCR duplicate or a optional duplicate
<br/>

## Checking out flag for some reads 

In [20]:
samfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")
i = 0
someReads = {6,124, 5789}
for read in samfile:
    if(i in someReads):
        print(f"Read name is {read.qname} and read flag is {read.flag}")
    i+=1

Read name is C0HVYACXX120402:3:1307:14105:43830 and read flag is 99
Read name is D0RE2ACXX120401:3:1107:20869:18508 and read flag is 147
Read name is D0RE2ACXX120401:3:2107:2597:112681 and read flag is 163


<br/>
<b> Flag for C0HVYACXX120402:3:1307:14105:43830 read is 99 =  0110 0011b </b> <br/>
1 - the read is paired in the sequencing, no metter whether it is mapped in pair <br/>
1 - the read is mapped in a proper pair <br/>
0 - not unmapped <br/>
0 - mate is not unamapped <br/>
0 - foward stand <br/>
1 - mate strand is negative <br/>
1 - the read is first read in pair <br/>
0 - the read is not second read in pair <br/>
<br/>

<br/>
<b> Flag for D0RE2ACXX120401:3:1107:20869:18508 read is 147 =  1001 0011b </b> <br/>
1 - the read is paired in the sequencing, no metter whether it is mapped in pair <br/>
1 - the read is mapped in a proper pair <br/>
0 - not unmapped <br/>
0 - mate is not unamapped <br/>
1 - reverse stand <br/>
0 - mate strand is not negative <br/>
0 - the read is not first read in pair <br/>
1 - the read is second read in pair <br/>
<br/>


<br/>
<b> Flag for D0RE2ACXX120401:3:2107:2597:112681 read is 163 =  1010 0011b </b> <br/>
1 - the read is paired in the sequencing, no metter whether it is mapped in pair <br/>
1 - the read is mapped in a proper pair <br/>
0 - not unmapped <br/>
0 - mate is not unamapped <br/>
0 - foward stand <br/>
1 - mate strand is negative <br/>
0 - the read is not first read in pair <br/>
1 - the read is second read in pair <br/>
<br/>

## Calculate

In [21]:
samfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

numUnmappedReads = 0 # number of unmapped reads
numOfReads = 0 #Total number of reads
readsMapQis0 = 0 #Number of reads with mapping quality 0
sumOfQuality = 0 #sum of mapping quality for all the reads

for read in samfile:
    numOfReads +=1
    sumOfQuality += read.mapq
    if(read.mapq == 0):
        readsMapQis0 += 1
    if(read.flag & 0x0004): 
        numUnmappedReads += 1

In [22]:
print(f"Total number of reads is {numOfReads}.")
print(f"In the file there are {numUnmappedReads} unmapped reads.")
print(f"Number of reads with mapping quality zero is {readsMapQis0}. ")
avg = sumOfQuality/numOfReads
print(f"Average mapping quality for all reads is {avg}.")
avg = sumOfQuality/(numOfReads-readsMapQis0)
print(f"Average mapping quality for reads if reads with 0 mapp quality are filttred out is {avg}.")

Total number of reads is 2921629.
In the file there are 17765 unmapped reads.
Number of reads with mapping quality zero is 126628. 
Average mapping quality for all reads is 55.91379158681681.
Average mapping quality for reads if reads with 0 mapp quality are filttred out is 58.446975510921106.


In [23]:
samfile.close()